### ISSUE TITLE
Output value instead of key from select list field items

### ISSUE THREAD

**Issue by timwood** [RESTful-Drupal/restful#718]
Is this possible? If so, is there an example of what to put in publicFieldsInfo function to get it to output the value instead of the key?

Thanks!

**Comment by e0ipso** [comment]
Could you add a bit more context to what you intend to do? I'm not sure I am following.

**Comment by timwood** [comment]
We have a field type of "List (text)" with the Select List widget and a few simple options stored as key -> value pairs. Where key is a numeric value 0, 1, 2, 3, 4, etc. and value is some text value such as "FOR IMMEDIATE RELEASE" or "AS PREPARED FOR DELIVERY". When I output the field in publicFieldsInfo function using 'property' => 'field_maching_name', the key is output in the API instead of the text value.

Does this help?

**Comment by e0ipso** [comment]
Have you looked into process callbacks?

**Comment by timwood** [comment]
I have not looked into process callbacks, but I'll research that. Thanks.

**Comment by jeff-h** [comment]
@timwood — https://github.com/RESTful-Drupal/restful/wiki/2.-Defining-the-public-fields-(1.x)#available-public-field-properties might be useful.

**Comment by timwood** [comment]
@jeff-h Thanks for pointing me to that page, I didn't reven realize this project had a Wiki! I'd previously reviewed the docs folder which points to this inline documentation about the callbacks with similar information. https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/plugins/restful/RestfulEntityBase.php#L47-L52

We have a few callbacks already, but I'm still trying to figure out how to get the select list item value instead of key. Will keep at it, thanks.

**Comment by jeff-h** [comment]
Yeah, sorry! The docs need consolidating :) It's my fault as I started pumping stuff into the wiki, but haven't had a chance to migrate the existing docs folder content into it.

Overall I'm finding it a little hard to understand what your question is, sorry! Could you post a copy of the request you're doing that isn't giving you the results you want? e.g.

`GET http://mydrupalsite.com/api/v1.0/myEntity/5`

or are you doing an "API discovery" call such as

`OPTIONS http://mydrupalsite.com/api` or perhaps `OPTIONS http://mydrupalsite.com/api/v1.0/myEntity`

and perhaps post a copy of the JSON response you're getting.

**Comment by timwood** [comment]
Just doing a regular GET call to an API for a node entity which has a field of type text (select list). So the field has five options stored as key/value pairs. When someone adds content (through the Drupal UI) they can select one option from the list. Restful outputs the KEY from the field instead of the value.

You should be able to replicate by adding a field to node of type text (select list) and adding some options for the field. Then output the field in the API. Does this make sense?

**Comment by timwood** [comment]
Drupal field config UI:
<img width="655" alt="screen shot 2015-11-04 at 5 35 03 pm" src="https://cloud.githubusercontent.com/assets/91050/10954288/8173c6ba-831a-11e5-9b61-8727e360ea30.png">

Drupal node add UI:
<img width="250" alt="screen shot 2015-11-04 at 5 31 25 pm" src="https://cloud.githubusercontent.com/assets/91050/10954241/4df3e504-831a-11e5-8464-0e23b4c715a7.png">

API output:
<img width="250" alt="screen shot 2015-11-04 at 5 33 55 pm" src="https://cloud.githubusercontent.com/assets/91050/10954246/5671e726-831a-11e5-8f95-f2f7f6098ca6.png">

**Comment by jeff-h** [comment]
That's super-clear now, thanks :D

My gut reaction is that what you're seeing is what it is supposed to do, but of course I can see your need for the "value" as well.

You could almost certainly just modify the `release_status` public field, or create a new one, and return the info you're after. See the [Defining the public fields](https://github.com/RESTful-Drupal/restful/wiki/2.-Defining-the-public-fields-%281.x%29) section in the wiki.

``` php
$public_fields['myField'] = array(
  'callback' => array($this, 'myFieldDataProviderCallbackWithALongName'),
);
```

where `myFieldDataProviderCallbackWithALongName` is a function you would write that would return the `value` instead of the key, for the field you want.

It would probably also be possible with RESTful's formatters, although I haven't used those yet.

**Comment by timwood** [comment]
It's the writing of the function that would return the value instead of the key which is currently my roadblock. I would have thought someone would have run into this but haven't found example code yet. I'll keep working on it and if I come up with something halfway decent I'll add it here or contribute it to the documentation Wiki. Thanks!

**Comment by jeff-h** [comment]
I think what you're hoping for might just be achieved by:

``` php
    $public_fields['release_status_value'] = array(
      'property' => 'field_release_status',
      'wrapper_method' => 'label',
    );
```

It quite likely may seem all a bit "magic", but the basis of this is that the `property` describes a field, which is wrapped using Drupal entity metadata wrappers, which provide a number of methods depending on the field. Thus, research into how to achieve this is not constrained to RESTful, but rather to the bigger picture of Drupal entity metadata wrappers.

The following answer describes how to access different aspects of a wrapped select field via wrapper methods:
http://drupal.stackexchange.com/questions/152397/entity-metadata-wrapper-how-to-get-field-value-not-key

By extension then, you could for example retrieve the entire set of options by doing `optionsList` as described in the last answer on the above url.

``` php
    $public_fields['release_status_options'] = array(
      'property' => 'field_release_status',
      'wrapper_method' => 'optionsList',
    );
```

**Comment by timwood** [comment]
@jeff-h You rock! That works perfectly.

Unfortunately, I realized changing the field output doesn't change the filter capability/functionality as I already discovered when trying to convert the date format from unixtime to ISO8601. Which is maybe why you suggested outputting the release status value as a different object (release_status_value) in the API?

**Comment by jeff-h** [comment]
Glad it worked :) You could probably modify the filter as per the [filter on term name](https://github.com/RESTful-Drupal/restful/wiki/Filter-on-term-name) example in the wiki.

Re: outputting release status value separately, I think it's good practice to use an id (the "key" in your key-value pairs) instead of the value because the value could change e.g. when a client decides to change your all-caps to title case, or when the site has to launch in a second language.

Thus, if you filter on release_status = 0 it will still work, even if key 0 ends up displayed as "For immediate release." or "For øjeblikkelig løsladelse" :p
